altair/utils/schemapi.py

```diff
@@ -486,6 +487,14 @@ def _subclasses(cls: type[Any]) -> Iterator[type[Any]]:
             yield cls
 
 
+def _from_array_like(obj: Iterable[Any], /) -> list[Any]:
+    try:
+        ser = nw.from_native(obj, strict=True, series_only=True)
+        return ser.to_list()
+    except TypeError:
+        return list(obj)
```

Comment:
Since anything that got here **must** be an `Iterable`, it will always be accepted by `list` https://docs.python.org/3/library/stdtypes.html#list

Reply:
Calling `list(iterable)` will defer to `iterable.__iter__`.

At least in the case of `pandas` and `polars`, calling `__iter__` vs `to_list` will perform a different operation.

`pandas`:
- [`__iter__`](https://github.com/pandas-dev/pandas/blob/63dc1bb4f99d24b46bacb113d740d54459fdbe5e/pandas/core/base.py#L873)
- [`to_list`](https://github.com/pandas-dev/pandas/blob/63dc1bb4f99d24b46bacb113d740d54459fdbe5e/pandas/core/base.py#L871)

`polars`:
- [`__iter__`](https://github.com/pola-rs/polars/blob/9978d88b15f3ea85195c36b1060d1d9f77369c93/py-polars/polars/series/series.py#L1224)
- [`to_list`](https://github.com/pola-rs/polars/blob/9978d88b15f3ea85195c36b1060d1d9f77369c93/py-polars/polars/series/series.py#L4036)

With the `polars` case, the simplest explanation is that explicitly asking for a `list` allows more of the operation to be performed in `rust`.

---

> which would make the code cleaner?

I'm not sure I agree on this part. 
I thought using `nw.from_native(obj, strict=True, series_only=True)` was:
- Very explicit in what is happening
- Defers a huge amount of type checking to an external library
- Ensures the most performant path to a `list` is taken
- Can potentially support even more types in the future, by simply changing the import to `import narwhals.stable.v2`

---

@joelostblom if you think this info would be helpful in the docstring for `_from_array_like` - I'd be happy to add it

Reply:
@MarcoGorelli I was wondering if there was a relation to `narwhals.Series.__iter__` not being defined yet? 
https://github.com/narwhals-dev/narwhals/blob/94eb1e229915bf0613ff92588332406c79959c9f/narwhals/series.py#L18

Also please correct me on any of my points in https://github.com/vega/altair/pull/3501#discussion_r1691248267 if you have more insight

Reply:
hey @dangotbanned - I think it's not necessary? At least, there's `__getitem_` and `__len__`

The rest of your points look good to me, esp. using the native `to_list`

> Can potentially support even more types in the future, by simply changing the import to import narwhals.stable.v2

Unless there's some breaking API change we just can't work around, any new library should still be available for `narwhals.stable.v1`

Reply:
Thanks @MarcoGorelli 

> I think it's not necessary? At least, there's `__getitem_` and `__len__`

To be clear I agree and I'm not asking for `__iter__` to be included.
We're not even using `narwhals.Series` directly here 😅